### ISSUE TITLE
Very slow gets of big values

### ISSUE THREAD

**Issue by qzb** [NodeRedis/node_redis#678]
I noticed that getting big values take more time than it should. Worse, it looks like time of get grows expotentially with size of value.

I wrote following script to test it:
```js
var redis = require('redis');
var fs = require('fs');

var key = Date.now().toString();
var data = fs.readFileSync('./data.bin'); // big file (25-100MB)

var client = redis.createClient(6379, 'localhost');

client.on('connect', function () {
    var time = Date.now();
    client.set(key, data, function (err) {
        console.log('set time: %s', Date.now() - time);
        time = Date.now();
        client.get(key, function (err, data) {
            console.log('get time: %s', Date.now() - time);
            process.exit();
        });
    });
});
```
When I pass to it 25MB of data, it takes about 6 seconds to get value from local redis, for 50MB it takes about 25 seconds and for 100MB it takes near 100 seconds.

**Comment by ajones513** [comment]
My results...

Using Hiredis | Size | Write (ms) | Read (ms)
---|---|---|---
No | 25 MB | 22 | 4114
No | 50 MB | 39 | 15128
Yes | 25 MB | 24 | 1026
Yes | 50 MB | 38 | 2007

Looks pretty linear with Hiredis.

**Comment by mranney** [comment]
Thanks for digging into this. I have confirmed these results, and there is not an easy way to fix this with the JS parser. It's certainly not impossible to fix in the JS parser, but it'll be some delicate surgery to do without slowing down or breaking the common case. I guess this is a good reason to keep hiredis working, at least until we can fix it.

For the curious, here's a flame graph of the test program above:

http://ranney.com/redis.svg

This program spends all of its time copying buffers around. hiredis wins because it calls JS back with a single 50MB result.

**Comment by brycebaril** [comment]
I've done a lot of benchmarking as well and have seen this is an issue with the parser. I've played with some things to try and fix it but it would be an api change for sure.

**Comment by b-long** [comment]
:+1:

**Comment by blainsmith** [comment]
In order to better support this project and its new group of collaborators under a new org we are trying to clean up and close issues older than 6 months. Your issue may have been fixed since the issue was open, however, if you are still experiencing a problem **please** re-open this issue and we will label it accordingly.

**Comment by BridgeAR** [comment]
Reopening as the js parser is still not handling big values well

**Comment by BridgeAR** [comment]
I analyzed the parser in depth and the cause of the problems seems to be the following:

If we have data bigger than 2^16 we'll have n chunks and the parser is going to concat n times creating a new buffer each time that increases in size and it's also parsing the whole buffer each time a new chunk is incoming (=> both leading to an increase in time per additional chunk, so it's exponential).
Also the stack might be getting bigger with each stack (each chunk will throw on top of each other) but I doubt that this is very substantial.

To fix the parser it has to be refactored in a way that we parse each chunk only once. This is likely not trivial though.

**Comment by qzb** [comment]
I think [rvagg/bl](https://github.com/rvagg/bl) module could be a solution for this problem.

**Comment by BridgeAR** [comment]
@qzb thx for the suggestion but this can't be used in our case. The stream is kept open all the time and we have no "end" besides closing the connection explicitly or losing the connection.

**Comment by qzb** [comment]
Maybe I'm missing something but I think I just fixed this bug using bl module: #900

**Comment by BridgeAR** [comment]
@qzb I found a fix that works ok for now (ff857f97aec48c74d16b13b4a07d6f00256c5b54) but I'm planing on rewriting the parser from scratch. The parser could be faster if rewritten and handling big values will be a lot faster than with this fix too. 

Would you be so kind and give the fix a try?

**Comment by BridgeAR** [comment]
Finally the issue is fixed :)

From now on the js parser handles big return values just as fast as the hiredis parser
